"Double Write" Identity Allocation Mode for seamless KVStore -> CRD identity migration

antonipp opened the pull request:
# Description
See detailed description of the problem in https://github.com/cilium/cilium/issues/30689

The second commit of this PR adds the new `--identity-allocation-mode=double-writeread{kvstore,crd}` mode described in the issue above to control whether the read operations happen from the KVStore or from CRDs. The main file to review is `pkg/kvstore/allocator/doublewrite/backend.go`.

The third commit of this PR adds a new Cilium Operator controller called "Double Write Metric Reporter" which helps with monitoring the identity migration progress and alerts about identity inconsistencies between KVStore and CRDs. It emits both logs and metrics. The main file to review is `operator/doublewrite/metricreporter.go`.

The migration process, as well as the reported metrics are documented in `Documentation/operations/upgrade.rst` and `Documentation/observability/metrics.rst`.

```release-note
Add "Double Write" Identity Allocation Mode for seamless KVStore <-> CRD identity migration
```

giorio94 commented on operator/identitygc/gc.go:
```diff
@@ -123,6 +123,16 @@ func registerGC(p params) {
+			case option.IdentityAllocationModeDoubleWrite:
+				crdErr := gc.startCRDModeGC(ctx)
+				if crdErr != nil {
+					log.WithError(crdErr).Error("Unable to start CRD mode GC")
```
I would return the error for symmetry with the cases above. Same for the kvstore GC below.

giorio94 commented on pkg/doublewrite/doublewritebackend.go:
```diff
@@ -0,0 +1,198 @@
+
+func (d *doubleWriteBackend) Encode(v string) string {
+	// Works for both CRD and etcd KVStore
+	return v
```
This works only assuming that the kvstore is etcd, as the `Encode` implementation in the consul client is different. I appreciate that consul support is mostly deprecated, but I'd still suggest adding a comment to outline that, and maybe consider the doublewrite allocator to be incompatible with it (hence raising an early error in that case), to prevent susprises.

antonipp replied:
It was a bit complicated to detect the KVStore implementation at runtime, so I simply added a comment and a callout in the docs

giorio94 commented on pkg/k8s/identitybackend/identity.go:
```diff
@@ -47,7 +47,7 @@ const (
 
 func NewCRDBackend(c CRDBackendConfiguration) (allocator.Backend, error) {
-	return &crdBackend{CRDBackendConfiguration: c}, nil
+	return &CRDBackend{CRDBackendConfiguration: c}, nil
```
Is exporting this struct really necessary? If I'm not missing anything, you are using interface methods only. And the same seems to apply to the KVStoreBackend one, expect that in that case you are also using the newly introduced `ListIDs` method, which could be abstracted through an interface.

antonipp replied:
Makes sense, I refactored the code to not export these structs

giorio94 commented on pkg/doublewrite/doublewritebackend.go:
```diff
@@ -0,0 +1,198 @@
+	if d.readFromKVStore {
+		return kvStoreKey, kvStoreErr
+	}
+	return crdKey, crdErr
```
Why is the error returned for a single backend only? I'd personally assume that we want to return it from both for all "write" operations, to ensure that we subsequently perform a retry even if only one failed for consistency. Am I missing anything? 

Another aspect to consider is how to ensure that these operations are atomic. Let's say that for instance the allocation succeeded in CRD mode, but then failed in kvstore mode. My understanding is that we should rollback the CRD allocation to remove the spurious CiliumIdentity object.

antonipp replied:
So I think trying to make these operations atomic would make the code exponentially more complex but wouldn't add a whole lot of benefit. 

You would need a system to rollback to a previous state if one of the writes failed: for object creations that would probably be possible with careful deletes but for object updates you would need to store the previous state in memory and also account for race conditions due to concurrent updates... IMO that's too complex, especially for a backend which is supposed to be used only for a short period of time.

Moreover, the current code already handles failed writes to the "primary" backend (i.e. the one used for reads) and there is alerting which helps with detecting discrepancies (in case a write to the "primary" backend succeeded but failed for the "secondary" backend).

giorio94 replied:
> So I think trying to make these operations atomic would make the code exponentially more complex but wouldn't add a whole lot of benefit.

Yeah, fair enough. I agree that it makes sense to skip that considering that it may not be possible to actually revert the previous changes, or lead to additional inconsistencies.

I still wonder whether it would be better to return an error if any of the two backend operations failed though, as that would allow the caller to detect that something went wrong, and perform a retry. My assumption is that it should help to automatically resolve temporary errors (e.g., kvstore blips, or creation conflicts), which may otherwise be retried only during a subsequent agent restart. WDYT?

antonipp replied:
I looked into it a bit more and I could technically refactor the code to return errors with something like this:
```diff
@@ -122,10 +126,12 @@ func (d *doubleWriteBackend) UpdateKey(ctx context.Context, id idpool.ID, key al
        crdErr := d.crdBackend.UpdateKey(ctx, id, key, true)
        if crdErr != nil {
                log.WithFields(logrus.Fields{logfields.Identity: id.String(), logfields.Key: key.String(), "reliablyMissing": reliablyMissing}).WithError(crdErr).Error("CRD backend failed to update key")
+               return crdErr
        }
        kvStoreErr := d.kvstoreBackend.UpdateKey(ctx, id, key, true)
        if kvStoreErr != nil {
                log.WithFields(logrus.Fields{logfields.Identity: id.String(), logfields.Key: key.String(), "reliablyMissing": reliablyMissing}).WithError(kvStoreErr).Error("KVStore backend failed to update key")
+               return kvStoreErr
        }
```
However, this adds a small complication: now if a CRD update fails, the kvstore update will never go through. My initial idea was that in `doublewrite-readkvstore` mode, the kvstore writes should be the "primary" ones and CRD writes are made on a best-effort basis, trying to not interfere with kvstore operations (and vice-versa in `doublewrite-readcrd` mode). This way the `doublewrite-readkvstore` functions in the same way as the regular `kvstore` mode but with best-effort CRD writes in parallel.

I could refactor the code even more to take this into account:
```go
func (d *doubleWriteBackend) UpdateKey(ctx context.Context, id idpool.ID, key allocator.AllocatorKey, reliablyMissing bool) error {
	if d.readFromKVStore {
		// If we are reading from the kvstore, we perform the kvstore update first
		kvStoreErr := d.kvstoreBackend.UpdateKey(ctx, id, key, true)
		if kvStoreErr != nil {
			log.WithFields(logrus.Fields{logfields.Identity: id.String(), logfields.Key: key.String(), "reliablyMissing": reliablyMissing}).WithError(kvStoreErr).Error("KVStore backend failed to update key")
			return kvStoreErr
		}
		crdErr := d.crdBackend.UpdateKey(ctx, id, key, true)
		if crdErr != nil {
			log.WithFields(logrus.Fields{logfields.Identity: id.String(), logfields.Key: key.String(), "reliablyMissing": reliablyMissing}).WithError(crdErr).Error("CRD backend failed to update key")
			return crdErr
		}
	}
	// If we are reading from CRDs, we perform the CRD update first
	crdErr := d.crdBackend.UpdateKey(ctx, id, key, true)
	if crdErr != nil {
		log.WithFields(logrus.Fields{logfields.Identity: id.String(), logfields.Key: key.String(), "reliablyMissing": reliablyMissing}).WithError(crdErr).Error("CRD backend failed to update key")
		return crdErr
	}
	kvStoreErr := d.kvstoreBackend.UpdateKey(ctx, id, key, true)
	if kvStoreErr != nil {
		log.WithFields(logrus.Fields{logfields.Identity: id.String(), logfields.Key: key.String(), "reliablyMissing": reliablyMissing}).WithError(kvStoreErr).Error("KVStore backend failed to update key")
		return kvStoreErr
	}
	return nil
}
```
But this makes the code more convoluted (I'll need to write the same logic for all write methods basically). What's your take on this?

giorio94 replied:
> I looked into it a bit more and I could technically refactor the code to return errors with something like this

Yeah, that is what I was thinking about when I wrote the above comment, so that the operation would be retried regardless of which backend failed. My main concern with only returning errors for the primary backend is that missing identities on the secondary one would cause divergences, that would be possibly fixed on agent restart only.

> However, this adds a small complication: now if a CRD update fails, the kvstore update will never go through.

Yes, but the operation would be retried, and eventually both backends should succeed. I agree that a persistent failure of one would also break the other, but that seems a pretty catastrophic situation to me.

> I could refactor the code even more to take this into account:

I agree that this looks pretty convoluted, and I don't think it would help much either, as, at least for a few methods, the caller would just retry with a different ID. Hence, there's not much point in ensuring that the primary update goes through first if the identity is discarded anyways in case of errors. 

In any case, I'd like to also get a second opinion of someone with more context on the internals of the allocator logic (@squeed maybe?).

squeed replied:
Because migration is expected to be reversible (as outlined in the original proposal), we must always write to both datastores, as @giorio94 observes.

This is especially critical given that different agents may have a different understanding of "primary". As configuration changes are rolled out, *if* we were to ignore secondary-side failures, we could have pretty serious divergence.

Given that most migrations are probably to the CRD store, and if the k8s apiserver goes down, nothing is happening anyways, I believe the right pattern is to write to CRD first, then kvstore. But, if kvstore fails, we must revert our CRD change and return failure.

There are still some nasty corner cases around interleaving failures between doublewrite and non-doublewrite agents, but I think those are very unlikely.

antonipp replied:
Makes sense, I wrote some code to handle this for the allocation functions: https://github.com/cilium/cilium/pull/31920/files#diff-ce8e3e1201058922bca6b6b8fd0ffc3a9ac595a34621246e05daf000f343bcb5R80-R100 
LMK what you think.

And also added a test: https://github.com/cilium/cilium/pull/31920/files#diff-08f67e47fb93b88a978cbf9b31e32f4af4b01ac36b1a78b9890201cf654b8d6aR89-R109

I didn't add the same logic for the `UpdateKey` / `AcquireReference` functions because rolling back the CRD update is not that important there: these functions basically update the heartbeat timestamp on the CRD object, so in the worst case, the GC of the identity will be delayed.

giorio94 commented on pkg/doublewrite/doublewritebackend.go:
```diff
@@ -0,0 +1,198 @@
+
+func (d *doubleWriteBackend) ListAndWatch(ctx context.Context, handler allocator.CacheMutations, stopChan chan struct{}) {
+	d.crdBackend.ListAndWatch(ctx, handler, stopChan)
+	d.kvstoreBackend.ListAndWatch(ctx, handler, stopChan)
```
This doesn't look correct to me, because the handler will be called twice for all the different identities, once when received from the informer and another time when received from the kvstore watcher. It will also likely cause a panic, as `OnListDone` would be called twice, hence causing a [channel to be closed twice](https://github.com/DataDog/cilium/blob/ai%2Fdouble-write-backend/pkg/allocator/cache.go#L123).

antonipp replied:
You are right, closing the channel was an issue, I refactored the code to account for that.
There is one subtlety: the `crdBackend` cannot do any read operations if `ListAndWatch` has not been called beforehand, because it's the only place where the underlying `Store` is initialized:
https://github.com/cilium/cilium/blob/797330cee40330c505fb97f6dd38446afb96ec78/pkg/k8s/identitybackend/identity.go#L369-L371
I didn't attempt to refactor that because it would be a non-trivial change. However, this forced me to call `ListAndWatch` in a number of places to force the initialization of the `Store`:
https://github.com/cilium/cilium/blob/f71bbf171602d800d63324db9d348717399c6b94/pkg/kvstore/doublewrite/backend.go#L198-L210
https://github.com/cilium/cilium/blob/240f36651260be0cec3d5ab0aab106cedc7180c8/operator/doublewrite/metricreporter.go#L86-L88

giorio94 commented on pkg/doublewrite/doublewritebackend_test.go:
New unit tests should be directly written using the testing framework rather than using `checkmate`: https://docs.cilium.io/en/stable/contributing/testing/unit/#migrating-tests-off-of-gopkg-in-check-v1

antonipp replied:
Refactored the code to use the new framework 👍

giorio94 commented on daemon/cmd/daemon_main.go:
```diff
@@ -422,6 +422,9 @@ func InitGlobalFlags(cmd *cobra.Command, vp *viper.Viper) {
 	option.BindEnv(vp, option.IdentityAllocationMode)
 
+	flags.Bool(option.IdentityAllocationModeDoubleWriteReadFromKVStore, true, "Whether to read from the KVStore when using the Double-Write identity allocation mode")
+	option.BindEnv(vp, option.IdentityAllocationModeDoubleWriteReadFromKVStore)
```
How about encoding this information as part of the identity allocation mode, rather than introducing a separate parameter? Something like `doublewrite-readcrd` and `doublewrite-readkvstore`?

antonipp replied:
Good idea, refactored the parameters to only accept `doublewrite-readcrd` and `doublewrite-readkvstore`

giorio94 commented on operator/identitygc/gc.go:
```diff
@@ -123,6 +123,16 @@ func registerGC(p params) {
+				if kvStoreErr != nil {
+					log.WithError(kvStoreErr).Error("Unable to start KVStore mode GC")
+				}
+				return nil
```
Please notice that CRD and kvstore GC work quite differently, which could be a likely source of temporary divergences (although they should eventually converge).

antonipp replied:
Yes, that makes sense, I noted this in the docs for end-users as well: https://github.com/cilium/cilium/pull/31920/files#diff-76c7d3d61bd3c6afeb8e936896c7a01046d530840abf499fd118e88e75d1a504R823

giorio94 reviewed (changes requested):
Nice work, thanks! This changes make migrating off the kvstore definitely way easier.

I did a first review iteration, and left a bunch of initial comments inline. 

Generally speaking, I'd kindly ask you to additionally:

* Extend the commit messages with additional details about the rationale of the changes. I appreciate that there's plenty of information in the PR description/linked issue, but having that as part of the commit messages as well would simplify understanding/bisecting the changes in future.
* Move preparatory refactoring changes into dedicated commits, as that makes it way easier to distinguish the changes that should be no ops from the ones actually introducing new features, and allow to better target the reviewers effort. Examples being the exporting of the CRD/kvstore backend structs (although that could be hopefully reverted), the introduction of the `KVStoreBackendConfiguration` struct, the preparatory log/metric changes, ...

squeed commented on install/kubernetes/cilium/values.yaml.tmpl:
```diff
@@ -1646,8 +1646,10 @@ hubble:
-# -- Method to use for identity allocation (`crd` or `kvstore`).
+# -- Method to use for identity allocation (`crd`, `kvstore` or `double-write`).
 identityAllocationMode: "crd"
+# -- Whether to read from the KVStore when using the Double-Write identity allocation mode (default true)
```
Rather than another parameter, what about modes `kvstore2crd` and `crd2kvstore`?

And yes, I know naming debates are tedious. I just try and avoid new Helm values where possible.

antonipp replied:
Yes, that's a good idea (similar to what Marco suggested yesterday: https://github.com/cilium/cilium/pull/31920#discussion_r1574442085)
I will refactor the code to have just one value.

ti-mo commented on pkg/doublewrite/doublewritebackend.go:
A few observations:
- Since this is kvstore-related, it should probably be a subpackage of `pkg/kvstore/`
- Why not call this file `backend.go`? The name stutters otherwise.
- If you do decide to leave it in `pkg/`, please add it to CODEOWNERS (I got pinged because of it)

antonipp replied:
Done, moved the code under `pkg/kvstore/` and renamed the file to `backend.go`

asauber reviewed (approved):
Config changes LGTM.

I wasn't sure if I liked the "coupling" of what is essentially two flags `enableDoubleWrites` and `readFromStore`. However, as implemented this coupling reduces potential misconfigurations.

giorio94 commented on operator/identitygc/crd_gc.go:
```diff
@@ -154,15 +154,15 @@ func (igc *GC) gc(ctx context.Context) error {
 	if ctx.Err() == nil {
-		igc.successfulRuns++
-		igc.metrics.IdentityGCRuns.WithLabelValues(LabelValueOutcomeSuccess).Set(float64(igc.successfulRuns))
+		igc.successfulRuns["crd"]++
```
Nit. Could use the `LabelIdentityTypeCRD` constant, rather than `"crd"`.

giorio94 commented on operator/cmd/kvstore_watchdog.go:
```diff
@@ -62,7 +62,13 @@ func getOldestLeases(lockPaths map[string]kvstore.Value) map[string]kvstore.Valu
+	backend, err := kvstoreallocator.NewKVStoreBackend(kvstoreallocator.KVStoreBackendConfiguration{
+		BasePath: cache.IdentitiesPath,
+		Suffix:   "",
+		Typ:      nil,
```
Nit. I'd omit the fields assigned to the default value.

giorio94 commented on operator/identitygc/kvstore_gc.go:
```diff
@@ -22,7 +22,7 @@ import (
 func (igc *GC) startKVStoreModeGC(ctx context.Context) error {
 	log.WithField(logfields.Interval, igc.gcInterval).Info("Starting kvstore identity garbage collector")
-	backend, err := kvstoreallocator.NewKVStoreBackend(cache.IdentitiesPath, "", nil, kvstore.Client())
+	backend, err := kvstoreallocator.NewKVStoreBackend(kvstoreallocator.KVStoreBackendConfiguration{BasePath: cache.IdentitiesPath, Suffix: "", Typ: nil, Backend: kvstore.Client()})
```
Nit. I'd omit the fields assigned to the default value.

giorio94 commented on operator/identitygc/kvstore_gc.go:
```diff
@@ -51,34 +51,34 @@ func (igc *GC) runKVStoreModeGC(ctx context.Context) error {
 
-			igc.failedRuns++
-			igc.metrics.IdentityGCRuns.WithLabelValues(LabelValueOutcomeFail).Set(float64(igc.failedRuns))
+			igc.failedRuns["kvstore"]++
```
Nit. Same here about the possibility of using `LabelIdentityTypeKVStore` instead of `"kvstore`".

giorio94 commented on pkg/kvstore/allocator/allocator.go:
```diff
@@ -554,9 +561,25 @@ func (k *kvstoreBackend) keyToID(key string) (id idpool.ID, err error) {
+func (k *kvstoreBackend) ListIDs(ctx context.Context) (identityIDs []idpool.ID, err error) {
+	identities, err := k.backend.ListPrefix(ctx, k.idPrefix)
+	if err != nil {
+		return []idpool.ID{}, err
```
Nit. I'd change `[]idpool.ID{}` to nil for consistency with the crd implementation. Same below.

giorio94 commented on install/kubernetes/cilium/templates/validate.yaml:
```diff
@@ -81,15 +81,15 @@
 {{/* validate clustermesh-apiserver */}}
 {{- if .Values.clustermesh.useAPIServer }}
-  {{- if ne .Values.identityAllocationMode "crd" }}
+  {{- if and (ne .Values.identityAllocationMode "crd") (ne .Values.identityAllocationMode "doublewrite-readkvstore") (ne .Values.identityAllocationMode "doublewrite-readcrd") }}
```
Hmm, I'd not prevent running the clustermesh-apiserver in combination with the migration settings, as that could be possibly useful for migrating environments with clustermesh enabled as well. For instance, one may enable the clustermesh-apiserver while performing the migration, and seamlessly switch the clustermesh configuration once the process has completed.

antonipp replied:
This is actually exactly what I am trying to do (i.e. allow the double write mode when clustermesh is enabled).
The logic will only block `identityAllocationMode` values different from `crd` and different from `doublewrite-readkvstore` and different from `doublewrite-readcrd`, which in our current state of things will basically only block the `kvstore` mode.

giorio94 replied:
Right, I misinterpreted that check.

giorio94 commented on install/kubernetes/cilium/values.yaml.tmpl:
```diff
@@ -1687,7 +1687,7 @@ hubble:
       - policy_denied
 
-# -- Method to use for identity allocation (`crd` or `kvstore`).
+# -- Method to use for identity allocation (`crd`, `kvstore` or `doublewrite-readkvstore` / `doublewrite-readcrd`).
```
I'd suggest to add the note that the "doublewrite" settings can be used for migration purposes here as well, as more likely to be seen by users compared to the rendered configmap.

giorio94 commented on pkg/kvstore/doublewrite/backend.go:
```diff
@@ -0,0 +1,225 @@
+
+func NewDoubleWriteBackend(c DoubleWriteBackendConfiguration) (allocator.Backend, error) {
+	crdBackend, _ := identitybackend.NewCRDBackend(c.CRDBackendConfiguration)
+	kvstoreBackend, _ := kvstoreallocator.NewKVStoreBackend(c.KVStoreBackendConfiguration)
```
Nit. It seems appropriate to check and propagate the errors here, although I agree that with the current implementation they would never occur.

giorio94 commented on pkg/kvstore/doublewrite/backend.go:
```diff
@@ -0,0 +1,225 @@
+
+type doubleWriteBackend struct {
+	crdBackend      *allocator.Backend
+	kvstoreBackend  *allocator.Backend
```
Are these actually required to be pointers, considering that `allocator.Backend` is an interface, and the underlying types are already pointers? That would allow dropping all the explicit de-referencing in the methods below.

giorio94 commented on pkg/kvstore/doublewrite/backend.go:
```diff
@@ -0,0 +1,225 @@
+}
+
+func (d *doubleWriteBackend) UpdateKey(ctx context.Context, id idpool.ID, key allocator.AllocatorKey, reliablyMissing bool) error {
+	crdErr := (*d.crdBackend).UpdateKey(ctx, id, key, true)
```
I guess the last parameter should be `reliablyMissing` rather than true. Same below.

antonipp replied:
It's a bit subtle but it is actually done on purpose... The idea is that when you first switch from `kvstore` to `doublewrite-readkvstore` mode, you want to automatically create CRD identities for kvstore identities which already exist. Forcing this to `true` here was the only solution I found for that. This will force the local key sync logic to create the CRD identities when looping through them here:
https://github.com/cilium/cilium/blob/e3e571c3c7896e5c46f36cd1b1d5b3f8de1ecdef/pkg/allocator/allocator.go#L885-L892

giorio94 replied:
OK, I think it would be definitely useful to get another pair of eyes from someone who knows better the internals of that logic (@squeed maybe?), but a couple of high level comments:

* It is definitely worth adding a comment to clarify the intent, as otherwise it is easy to miss, and it does look an oversight.
* I'm not entirely sure of the consequences, but that flag definitely modifies the behavior of the downstream methods. As for CRDs, it eventually triggers the creation of the corresponding CID, which would fail if already present. However, that case seems to be [short-circuited via AcquireReference](https://github.com/cilium/cilium/blob/74a025c60a9467cb590fc28c848460e05e8513c6/pkg/k8s/identitybackend/identity.go#L178-L185). As for KVStore, instead, it [determines](https://github.com/cilium/cilium/blob/74a025c60a9467cb590fc28c848460e05e8513c6/pkg/kvstore/allocator/allocator.go#L291-L295) whether it calls `CreateOnly` or `UpdateIfDifferent` (with the latter actually either creating or updating the key). With that in mind, it may be sensible to propagate the correct value to the kvstore backend, and only force it to be true for the CRD case.

antonipp replied:
I only added a comment for now but I agree that it's not a trivial problem and it would be great to get another opinion of this.
The only comment I have for now is that the solution to "only force it to be true for the CRD case" would not entirely suffice, because ideally we would want to have the possibility to revert from CRD mode back to KVStore mode as well, so we'd need to support both migration paths (KVStore -> CRD and CRD -> KVStore)

giorio94 replied:
> The only comment I have for now is that the solution to "only force it to be true for the CRD case" would not entirely suffice, because ideally we would want to have the possibility to revert from CRD mode back to KVStore mode as well, so we'd need to support both migration paths (KVStore -> CRD and CRD -> KVStore)

Sure. However, unless I'm missing something, in the kvstore backend, `reliablyMissing` only changes one operation from using `CreateOnly` (when true) to `UpdateIfDifferent` (when false). However,  `UpdateIfDifferent` is actually more general, as it also performs a creation if not present. Hence, forcing to to true should not be necessary there.

antonipp replied:
Oh I see, I read your initial message too quickly, this makes sense 👍

giorio94 commented on pkg/kvstore/doublewrite/backend.go:
```diff
@@ -0,0 +1,225 @@
+func (f NoOpHandler) OnModify(id idpool.ID, key allocator.AllocatorKey) {}
+func (f NoOpHandler) OnDelete(id idpool.ID, key allocator.AllocatorKey) {}
+
+func (d *doubleWriteBackend) ListAndWatch(ctx context.Context, handler allocator.CacheMutations, stopChan chan struct{}) {
```
`ListAndWatch` is expected to block until termination. Breaking that expectation would [break the consumers](https://github.com/cilium/cilium/blob/84594f6e4791a4bc454d3e79c3d3a9d18976210b/pkg/allocator/cache.go#L242-L245).

giorio94 commented on pkg/kvstore/doublewrite/backend.go:
```diff
@@ -0,0 +1,225 @@
+		go (*d.kvstoreBackend).ListAndWatch(ctx, handler, stopChan)
+		// We still need to run ListAndWatch for the CRD backend to initialize the underlying store.
+		// Since we don't need to use the results of the list operation, we can use a no-op handler
+		crdStopChan := make(chan struct{})
```
You could just pass `stopChan` to both, as the channel is only closed.

giorio94 commented on pkg/kvstore/doublewrite/backend_test.go:
```diff
@@ -0,0 +1,120 @@
+	"github.com/cilium/cilium/pkg/testutils"
+)
+
+func setup(tb testing.TB) (string, *k8sClient.FakeClientset, *allocator.Backend) {
```
Nit. return `allocator.Backend` by value, to drop the explicit de-referencing below.

giorio94 commented on pkg/kvstore/doublewrite/backend_test.go:
```diff
@@ -0,0 +1,120 @@
+	ids, err := kubeClient.CiliumV2().CiliumIdentities().List(context.Background(), metav1.ListOptions{})
+	require.Nil(t, err)
+	require.Len(t, ids.Items, 1)
+	require.Equal(t, ids.Items[0].Name, identityID.String())
```
Nit. The parameters should be inverted, as the signature is `t, expected, actual`.

giorio94 commented on pkg/kvstore/doublewrite/backend_test.go:
```diff
@@ -0,0 +1,120 @@
+	k := &key.GlobalIdentity{LabelArray: lbls.LabelArray()}
+	identityID := idpool.ID(10)
+	_, err := (*backend).AllocateID(context.Background(), identityID, k)
+	require.Nil(t, err)
```
Nit. `require.NoError`, which is a bit more idiomatic.

giorio94 commented on pkg/kvstore/doublewrite/backend_test.go:
```diff
@@ -0,0 +1,120 @@
+	require.Equal(t, ids.Items[0].Name, identityID.String())
+	require.EqualValues(t,
+		ids.Items[0].SecurityLabels,
+		map[string]string{fmt.Sprintf("%s:%s", lbls.LabelArray()[0].Source, lbls.LabelArray()[0].Key): lbls.LabelArray()[0].Value},
```
Nit. The parameters should be inverted, as the signature is `t, expected, actual`.

giorio94 replied:
Nit. `k.GetAsMap()`

giorio94 commented on pkg/kvstore/doublewrite/backend_test.go:
```diff
@@ -0,0 +1,120 @@
+	require.Len(t, kvPairs, 1)
+	require.Equal(t,
+		string(kvPairs[fmt.Sprintf("%s/id/%s", kvstorePrefix, identityID)].Data),
+		fmt.Sprintf("%s:%s=%s;", lbls.LabelArray()[0].Source, lbls.LabelArray()[0].Key, lbls.LabelArray()[0].Value),
```
Nit. `k.GetKey()`

giorio94 commented on pkg/kvstore/doublewrite/backend.go:
I'd suggest moving these files under `pkg/kvstore/allocator/doublewrite/`, being allocator related.

giorio94 commented on operator/doublewrite/cell.go:
```diff
@@ -0,0 +1,42 @@
+	"double-write-metric-reporter",
+	"Double-Write Metric Reporter",
+
+	cell.Config(Config{}),
```
Nit. Typically default parameters are implemented specifying `defaultConfig` here, and using `def.Interval` in the `Flags` method below. That allows further flexibility in case different modules needed to configure different defaults.

giorio94 commented on operator/doublewrite/metricreporter.go:
```diff
@@ -0,0 +1,159 @@
+	g.crdBackend = crdBackend
+	// Initialize the CRD backend store
+	g.crdBackendWatcherStopChan = make(chan struct{})
+	go g.crdBackend.ListAndWatch(context.Background(), doublewrite.NoOpHandler{}, g.crdBackendWatcherStopChan)
```
I'd suggest adding a waitgroup to ensure that this goroutine terminated when `Stop()` is called. In general, cells should ensure that no goroutines are still running after that the `Stop` method completes, to avoid violating the dependency-based start/stop order.

giorio94 commented on operator/doublewrite/metrics.go:
```diff
@@ -0,0 +1,55 @@
+	return &Metrics{
+		IdentityCRDTotalCount: metric.NewGauge(metric.GaugeOpts{
+			Namespace: metrics.CiliumOperatorNamespace,
+			Name:      "identity_crd_total_count",
```
I wonder if it could make sense to include `doublewrite` somewhere in the name (maybe `identity_doublewrite_crd_total_count`) to explicit the fact that these metrics are output only in doublewrite mode.

giorio94 commented on Documentation/cmdref/cilium-operator-alibabacloud.md:
These autogenerated changes should be moved to the second and third commit, as appropriate.

giorio94 commented on Documentation/observability/metrics.rst:
```diff
@@ -691,6 +691,30 @@ Name                                           Labels
+``identity_crd_total_count``             Disabled   The total number of CRD identities
+``identity_kvstore_total_count``         Disabled   The total number of identities in the KVStore
+``identity_crd_only_count``              Disabled   The number of CRD identities not present in the KVStore
+``identity_kvstore_only_count``          Disabled   The number of identities in the KVStore not present as a CRD
```
The default should be `Enabled` here, in the sense that the metrics are enabled by default (you didn't specify `Disabled: true` as part of the definition. Then they are obviously present only if the allocation mode is doublewrite.

giorio94 commented on Documentation/operations/upgrade.rst:
```diff
@@ -692,36 +692,48 @@ As the :term:`ConfigMap` is successfully upgraded we can start upgrading Cilium
+
+If these limitations are not acceptable, it is recommended to use the ":ref:`Double Write <double_write_migration>`" identity allocation mode instead.
+
 The steps below show an example of such a migration. It is safe to re-run the
```
I'd suggest to change "such a migration" with something more explicit (maybe the "one-off migration"?), as possibly confusing now with the additions above.

giorio94 commented on Documentation/operations/upgrade.rst:
```diff
@@ -790,14 +802,60 @@ Once the migration is complete, confirm the endpoint identities match by listing
+.. _double_write_migration:
+
+Migration with the "Double Write" identity allocation mode
+##########################################################
```
I'd suggest adding a note that this is a beta feature, and users may need extra care. This can be done with:

```
.. include:: ../beta.rst
```

giorio94 commented on operator/doublewrite/metricreporter.go:
```diff
@@ -0,0 +1,159 @@
+
+func (g *DoubleWriteMetricReporter) compareCRDAndKVStoreIdentities(ctx context.Context) error {
+	// Get CRD identities
+	crdIdentityIds, err := g.crdBackend.ListIDs(ctx)
```
I'd wait for synchronization to complete before retrieving the data from the store, as it would otherwise report unnecessary spikes of mismatches until identities have been synchronized from k8s.

giorio94 reviewed (changes requested):
Thanks! I did a second pass, and overall the logic seems pretty sound to me. I've left a bunch of comments inline -- they are quite a few, but mostly nits, except for a couple.

I'm still planning on having a final look and doing a bit of manual testing before approving the PR.

antonipp committed:
Refactoring to prepare the Double-Write id allocation mode implementation

- Create KVStoreBackendConfiguration in order to be consistent with CRDBackendConfiguration
- Add new method ListIDs() to the allocator.Backend interface
- Add new labels to Operator GC metrics in order to distinguish between KVStore and CRD GC data points
- Make logs clearer so that it's easier to distinguish between KVStore and CRD operations

Signed-off-by: Anton Ippolitov <anton.ippolitov@datadoghq.com>

antonipp committed:
Create the Double Write identity allocation backend

This backend allows Cilium to allocate identities as KVStore values *and* as CRDs at the same time. Its main purpose is to allow seamless migrations from KVStore to CRD identity allocation mode.
The backend has two versions: `doublewrite-readkvstore` and `doublewrite-readcrd`. These control whether identity read operations happen from the KVStore or from CRDs.

Signed-off-by: Anton Ippolitov <anton.ippolitov@datadoghq.com>

antonipp committed:
Create the Double Write Metric Reporter

When the "Double Write" identity allocation mode is enabled, the Double Write Metric Reporter will emit additional metrics and logs which help with monitoring the migration progress
and can be used for alerting about identity inconsistencies between KVStore and CRDs.

The Metric Reporter runs in the Cilium Operator.

Signed-off-by: Anton Ippolitov <anton.ippolitov@datadoghq.com>

antonipp committed:
Add Helm chart support for the Double Write identity allocation mode

Signed-off-by: Anton Ippolitov <anton.ippolitov@datadoghq.com>

antonipp committed:
Add documentation for the Double Write identity allocation mode

Signed-off-by: Anton Ippolitov <anton.ippolitov@datadoghq.com>

giorio94 commented on operator/identitygc/crd_gc.go:
```diff
@@ -168,15 +168,15 @@ func (igc *GC) gc(ctx context.Context) error {
 	if ctx.Err() == nil {
-		igc.successfulRuns++
-		igc.metrics.IdentityGCRuns.WithLabelValues(LabelValueOutcomeSuccess).Set(float64(igc.successfulRuns))
+		igc.successfulRuns[LabelIdentityTypeCRD]++
```
Given that the subsequent commit introduces the possibility for both GCs to run in parallel, access to this field (and the equivalent one about failures) shall be synchronized. However, it seems that they are only ever used to update the metrics, and they could be dropped in favor of just calling `Inc()` rather than `Set`.

antonipp replied:
Makes sense, I just simplified this to use `Inc()`

giorio94 commented on pkg/k8s/identitybackend/identity.go:
```diff
@@ -350,6 +350,21 @@ func getIdentitiesByKeyFunc(keyFunc func(map[string]string) allocator.AllocatorK
+		idParsed, err := strconv.ParseUint(identity.(*v2.CiliumIdentity).Name, 10, 64)
+		if err != nil {
+			return nil, err
+		}
```
I wonder if it wouldn't be better to skip incorrect identities (while printing a warning log), rather than returning an error. While that should never happen, it would prevent an incorrectly named identity from completely breaking this method. Same for the equivalent kvstore one.

giorio94 commented on Documentation/operations/upgrade.rst:
```diff
@@ -765,36 +765,48 @@ As the :term:`ConfigMap` is successfully upgraded we can start upgrading Cilium
-command if desired. It will identify already allocated identities or ones that
+There are two ways to achieve this: you can either run a one-off ``cilium preflight migrate-identity`` script
+which will perform a point-in-time copy of all identities from the kvstore to CRDs (added in Cilium 1.6), or use the "Double Write" identity
+allocation mode which will have Cilium manage identities in both the kvstore and CRD at the same time for a seamless migration (added in Cilium 1.16).
```
`s/1.16/1.17/`

giorio94 commented on operator/doublewrite/metricreporter.go:
```diff
@@ -0,0 +1,180 @@
+
+func (g *DoubleWriteMetricReporter) compareCRDAndKVStoreIdentities(ctx context.Context) error {
+	// Get CRD identities
+	<-g.crdBackendListDone
```
Nit. I'd select for `ctx.Done()` as well, to unblock in case the cell gets stopped before synchronization.

giorio94 commented on operator/doublewrite/metricreporter.go:
```diff
@@ -0,0 +1,180 @@
+		"KVStore identities: %d\n"+
+		"Identities only in CRD: %d. Example IDs (capped at %d): %v\n"+
+		"Identities only in KVStore: %d. Example IDs (capped at %d): %v\n",
+		len(crdIdentityIds), len(kvstoreIdentityIds), onlyInCrdCount, maxPrintedDiffIDs, onlyInCrdSample, onlyInKVStoreCount, maxPrintedDiffIDs, onlyInKVStoreSample)
```
New lines don't play well with the logger (at least the default one). The different information should be provided via `WithFields`. It could also make sense to print a different message in case everything is synchronized, to immediately identify that situation.

giorio94 commented on pkg/kvstore/allocator/doublewrite/backend.go:
```diff
@@ -0,0 +1,223 @@
+	return crdKey, crdErr
+}
+
+func (d *doubleWriteBackend) AcquireReference(ctx context.Context, id idpool.ID, key allocator.AllocatorKey, lock kvstore.KVLocker) error {
```
I did a bit of testing about the first part of the migration (kvstore -> doublewrite-readkvstore), and the agent reported both errors and warnings for all (or almost all) existing identities, along the lines of:

```
time="2024-07-01T13:34:55Z" level=error msg="CRD backend failed to acquire reference with lock" error="identity (id:\"41918\",key:\"[k8s:app=podinfo22 k8s:io.cilium.k8s.namespace.labels.kubernetes.io/metadata.name=default k8s:io.cilium.k8s.policy.cluster=kind-kind k8s:io.cilium.k8s.policy.serviceaccount=default k8s:io.kubernetes.pod.namespace=default]\") does not exist" identity=41918 key="[k8s:app=podinfo22 k8s:io.cilium.k8s.namespace.labels.kubernetes.io/metadata.name=default k8s:io.cilium.k8s.policy.cluster=kind-kind k8s:io.cilium.k8s.policy.serviceaccount=default k8s:io.kubernetes.pod.namespace=default]" subsys=double-write-allocator
time="2024-07-01T13:40:08Z" level=warning msg="Unable update CRD identity information with a reference for this node" error="identity (id:\"31737\",key:\"[k8s:app=podinfo34 k8s:io.cilium.k8s.namespace.labels.kubernetes.io/metadata.name=default k8s:io.cilium.k8s.policy.cluster=kind-kind k8s:io.cilium.k8s.policy.serviceaccount=default k8s:io.kubernetes.pod.namespace=default]\") does not exist" identity=31737 labels="[k8s:app=podinfo34 k8s:io.cilium.k8s.namespace.labels.kubernetes.io/metadata.name=default k8s:io.cilium.k8s.policy.cluster=kind-kind k8s:io.cilium.k8s.policy.serviceaccount=default k8s:io.kubernetes.pod.namespace=default]" subsys=crd-allocator
```

Eventually, the operator reported all identities are synchronized. However, just looking at the agent logs, it appears that the migration didn't work correctly. This aspect should be improved to provide clearer information to users (I guess there may be a race condition during the initial recreation). I've seen a bunch of similar errors on agent restart as well (only once though), which makes me think it may be somehow related with store synchronization.

antonipp replied:
You are right, I noticed these errors as well, I'll try to see if I can eliminate them somehow 🤔 It did look like it was due to a race condition so not sure how easy it would be to fix that

antonipp replied:
I spent a bit of time on this issue with error logs and I think I got the root cause: it happens when the agent starts for the first time with the `--identity-allocation-mode=doublewrite-readkvstore` setting and tries to retrieve IDs for identities which already exist. 

The agent first tries to retrieve / allocate the identity IDs for identity labels (keys) by calling [`Allocator.Allocate()`](https://github.com/cilium/cilium/blob/e86a4145cd19589e82f88efd33505fddf1e6c0b9/pkg/allocator/allocator.go#L643). 
The logic [checks](https://github.com/cilium/cilium/blob/722d00fbc075ad3bc20bdcd79fefc2f5470f1b98/pkg/allocator/allocator.go#L664-L675) whether the key is already in `localKeys`, which on first start-up is not the case, so it goes [further](https://github.com/cilium/cilium/blob/722d00fbc075ad3bc20bdcd79fefc2f5470f1b98/pkg/allocator/allocator.go#L678) to call [`lockedAllocate()`](https://github.com/cilium/cilium/blob/e86a4145cd19589e82f88efd33505fddf1e6c0b9/pkg/allocator/allocator.go#L49).

The `lockedAllocate()` code verifies whether the key already exists in the identity backend by [calling](https://github.com/cilium/cilium/blob/e86a4145cd19589e82f88efd33505fddf1e6c0b9/pkg/allocator/allocator.go#L510-L512) `GetIfLocked()`. In the `doublewrite-readkvstore` mode, this will only query the KVStore and will return some `value` because the key already exists. It means we enter [this branch](https://github.com/cilium/cilium/blob/722d00fbc075ad3bc20bdcd79fefc2f5470f1b98/pkg/allocator/allocator.go#L530-L541) and [reserve](https://github.com/cilium/cilium/blob/e86a4145cd19589e82f88efd33505fddf1e6c0b9/pkg/allocator/allocator.go#L540) a new local key (`firstUse` is `true` as well in this scenario).

We then [proceed to call](https://github.com/cilium/cilium/blob/e86a4145cd19589e82f88efd33505fddf1e6c0b9/pkg/allocator/allocator.go#L549) `a.backend.AcquireReference` which will call both the KVStore and the CRD `AcquireReference()` implementations. This is where the problem happens. For the KVStore there is no problem, it simply creates a `/value/<key>/<node>` KVStore entry to record that the key is used by the current node. However, the `AcquireReference()` call fails for the CRD backend because the actual CRD identity object has not been created yet! Basically all of [these checks](https://github.com/cilium/cilium/blob/722d00fbc075ad3bc20bdcd79fefc2f5470f1b98/pkg/k8s/identitybackend/identity.go#L139-L150) fail and the "identity does not exist" error is logged.

It's not really a race condition because the `localKeys` need to be properly populated first and they will then be used to create the CRD identities through the [`syncLocalKeys()`](https://github.com/cilium/cilium/blob/722d00fbc075ad3bc20bdcd79fefc2f5470f1b98/pkg/allocator/allocator.go#L860-L878) function later. (See https://github.com/cilium/cilium/pull/31920#discussion_r1650610739)

I don't see an easy option to suppress these logs 🤔 Maybe some additional logic in `doubleWriteBackend.AcquireReference()` which would check whether `crdErr` is "identity does not exist" and `kvStoreErr` is `nil`, and would not log the `crdErr`? (feels a bit ugly)
Another option could be having the Operator create missing CRD objects on the first switch to `--identity-allocation-mode=doublewrite-readkvstore`. So it wouldn't only monitor the CRD / KVStore inconsistencies but also fix them. However this feels tricky with the different GC schedules and would also entail a huge refactoring.

squeed replied:
You know, it's a bit ugly to check the error type and elide logging it, but it's totally reasonable. I honestly prefer a bit of ugly code that kicks down an error to, say, `.Debug()` if it meets a benign condition. People spend 10x more time reading logs than they do code.

antonipp replied:
Sure, I can add some logic to do that then 👍

antonipp replied:
Added it here: https://github.com/cilium/cilium/pull/31920/files#diff-ce8e3e1201058922bca6b6b8fd0ffc3a9ac595a34621246e05daf000f343bcb5R126-R135

giorio94 replied:
Thanks, makes sense to me.  Just an additional note that `syncLocalKeys` will still emit a warning log along the lines of:

> time="2024-07-25T08:58:37Z" level=warning msg="Unable update CRD identity information with a reference for this node" error="identity ..." subsys=crd-allocator

when `UpdateKey` is called, although the key is subsequently recreated as `reliablyMissing` is set to true.

However, it looks acceptable to me, considering that it only happens during the initial migration (the only easy fix I can think of would be lowering the severity of that log as well in the `doublewrite-readkvstore` mode).

giorio94 reviewed (changes requested):
Thanks for the changes. I've only a few remaining comments inline, which are mostly minor. The main aspect IMO would be addressing the error logs triggered when creating the CRD identities for the first time, as most likely confusing for users. 

Are you planning on adding an E2E tests covering the kvstore to CRD migration as well (also in a separate PR)? That would be definitely useful to prevent unintentionally breaking the doublewrite logic in the future.

joestringer commented on Documentation/operations/upgrade.rst:
```diff
@@ -820,7 +832,7 @@ The cilium preflight manifest requires etcd support and can be built with:
 
 Example migration
-~~~~~~~~~~~~~~~~~
+$$$$$$$$$$$$$$$$$
```
In Cilium docs we use `~` for this level of header. Please follow the same (same feedback below):

```suggestion
~~~~~~~~~~~~~~~~~
```

ysksuzuki reviewed (approved):
The Egress Gateway change LGTM.

Currently, Egress Gateway doesn't support Identity allocation mode `kvstore`. So the same applies to `doublewrite-readkvstore` and `doublewrite-readcrd`. Once users finish the migration and switch to `crd`, they can start using Egress Gateway.

learnitall commented on Documentation/operations/upgrade.rst:
```diff
@@ -463,36 +463,48 @@ As the :term:`ConfigMap` is successfully upgraded we can start upgrading Cilium
+The ``cilium preflight migrate-identity`` script is a one-off tool that can be used to copy identities from the kvstore into CRDs.
+It has a couple of limitations:
+
+* If any identity is created in the kvstore after the one-off migration has been completed, it will not be copied into CRDs.
```
```suggestion
* If an identity is created in the kvstore after the one-off migration has been completed, it will not be copied into a CRD.
```

learnitall commented on Documentation/operations/upgrade.rst:
```diff
@@ -463,36 +463,48 @@ As the :term:`ConfigMap` is successfully upgraded we can start upgrading Cilium
+* If any identity is created in the kvstore after the one-off migration has been completed, it will not be copied into CRDs.
+  This means that you need to perform the migration on a cluster with no identity churn.
+* There is no easy way to revert back to ``--identity-allocation-mode=kvstore`` if something goes wrong after
+  the Agents have been migrated to ``--identity-allocation-mode=crd``
```
```suggestion
  Cilium has been migrated to ``--identity-allocation-mode=crd``.
```

learnitall commented on Documentation/operations/upgrade.rst:
```diff
@@ -463,36 +463,48 @@ As the :term:`ConfigMap` is successfully upgraded we can start upgrading Cilium
+
+If these limitations are not acceptable, it is recommended to use the ":ref:`Double Write <double_write_migration>`" identity allocation mode instead.
+
+The steps below show an example of the migration using the ``cilium preflight migrate-identity`` script.
```
```suggestion
The following steps show an example of performing the migration using the ``cilium preflight migrate-identity`` script.
```

learnitall commented on Documentation/operations/upgrade.rst:
```diff
@@ -463,36 +463,48 @@ As the :term:`ConfigMap` is successfully upgraded we can start upgrading Cilium
 
 The steps below assume a stable cluster with no new identities created during
-the rollout. Once a cilium using CRD-backed identities is running, it may begin
+the rollout. Once a Cilium using CRD-backed identities is running, it may begin
```
```suggestion
the rollout. Once Cilium using CRD-backed identities is running, it may begin
```

learnitall commented on Documentation/operations/upgrade.rst:
```diff
@@ -563,12 +575,60 @@ Once the migration is complete, confirm the endpoint identities match by listing
+
+The "Double Write" Identity Allocation Mode allows Cilium to allocate identities as KVStore values *and* as CRDs at the
+same time. This mode also has two versions (``--identity-allocation-mode=doublewrite-readkvstore`` and ``--identity-allocation-mode=doublewrite-readcrd``)
+which allow controlling whether reads happen from KVStore or from CRDs.
```
```suggestion
same time. This mode also has two versions: one where the source of truth comes from the kvstore, and one where the source of truth comes from CRDs .
```

learnitall commented on Documentation/operations/upgrade.rst:
```diff
@@ -563,12 +575,60 @@ Once the migration is complete, confirm the endpoint identities match by listing
+
+.. note::
+
+    The "Double Write" mode is not compatible with Consul as the KVStore
```
```suggestion
    "Double Write" mode is not compatible with Consul as the KVStore
```

learnitall commented on Documentation/operations/upgrade.rst:
```diff
@@ -563,12 +575,60 @@ Once the migration is complete, confirm the endpoint identities match by listing
+   pure KVStore mode with the only difference being that all identities are duplicated as CRDs but are not used.
+#. Switch Cilium to “Double Write” mode with all reads happening from CRDs. This is equivalent to Cilium running in
+   pure CRD mode but identities will still be updated in the KVStore for possibility of fast rollback
+#. Switch Cilium to CRD mode. The KVStore will no longer be used and will be ready for decommission
```
```suggestion

#. Starting state: Cilium is running in KVStore mode.
#. Switch Cilium to “Double Write” mode with all reads happening from the KVStore. This is almost the same as the
   pure KVStore mode with the only difference being that all identities are duplicated as CRDs but are not used.
#. Switch Cilium to “Double Write” mode with all reads happening from CRDs. This is equivalent to Cilium running in
   pure CRD mode but identities will still be updated in the KVStore to allow for the possibility of a fast rollback.
#. Switch Cilium to CRD mode. The KVStore will no longer be used and will be ready for decommission.
```

learnitall commented on Documentation/operations/upgrade.rst:
```diff
@@ -563,12 +575,60 @@ Once the migration is complete, confirm the endpoint identities match by listing
+   pure CRD mode but identities will still be updated in the KVStore for possibility of fast rollback
+#. Switch Cilium to CRD mode. The KVStore will no longer be used and will be ready for decommission
+
+This will allow you to perform a gradual and seamless migration with possibility of fast rollback at steps 2 and 3.
```
```suggestion
This will allow you to perform a gradual and seamless migration with the possibility of a fast rollback at steps two or three.
```

learnitall commented on Documentation/operations/upgrade.rst:
```diff
@@ -563,12 +575,60 @@ Once the migration is complete, confirm the endpoint identities match by listing
+This will allow you to perform a gradual and seamless migration with possibility of fast rollback at steps 2 and 3.
+
+Furthermore, when the "Double Write" mode is enabled, the Operator will emit additional metrics which help with monitoring the
+migration progress and can be used for alerting about identity inconsistencies between KVStore and CRDs.
```
```suggestion
Furthermore, when the "Double Write" mode is enabled, the Operator will emit additional metrics to help monitor the
migration progress. These metrics can be used for alerting about identity inconsistencies between the KVStore and CRDs.
```

learnitall commented on Documentation/operations/upgrade.rst:
```diff
@@ -563,12 +575,60 @@ Once the migration is complete, confirm the endpoint identities match by listing
+#. Monitor the Operator metrics and logs to ensure that all identities have converged between the KVStore and CRDs. The relevant metrics emitted by the Operator are:
+
+   * ``cilium_operator_identity_crd_total_count`` and ``cilium_operator_identity_kvstore_total_count`` reporting the total number of identities in CRDs and KVStore respectively.
+   * ``cilium_operator_identity_crd_only_count`` and ``cilium_operator_identity_kvstore_only_count`` reporting the number of
```
```suggestion
   * ``cilium_operator_identity_crd_total_count`` and ``cilium_operator_identity_kvstore_total_count`` report the total number of identities in CRDs and KVStore respectively.
   * ``cilium_operator_identity_crd_only_count`` and ``cilium_operator_identity_kvstore_only_count`` report the number of
```

learnitall commented on Documentation/operations/upgrade.rst:
```diff
@@ -563,12 +575,60 @@ Once the migration is complete, confirm the endpoint identities match by listing
+
+   In case further investigation is needed, the Operator logs will contain detailed information about the discrepancies between KVStore and CRD identities.
+   Note that Garbage Collection for KVStore identities and CRD identities happens at slightly different times, so it is possible to see some discrepancies in the metrics
+   for some periods of time (depending on ``--identity-gc-interval`` and ``--identity-heartbeat-timeout`` settings)
```
```suggestion
   Note that Garbage Collection for KVStore identities and CRD identities happens at slightly different times, so it is possible to see discrepancies in the metrics
   for certain periods of time, depending on ``--identity-gc-interval`` and ``--identity-heartbeat-timeout`` settings.
```

learnitall commented on Documentation/operations/upgrade.rst:
```diff
@@ -563,12 +575,60 @@ Once the migration is complete, confirm the endpoint identities match by listing
+   Note that Garbage Collection for KVStore identities and CRD identities happens at slightly different times, so it is possible to see some discrepancies in the metrics
+   for some periods of time (depending on ``--identity-gc-interval`` and ``--identity-heartbeat-timeout`` settings)
+#. Once all identities have converged, re-deploy the Operator and the Agents with ``--identity-allocation-mode=doublewrite-readcrd``.
+   This will make Cilium read identities only from CRDs but still write them to KVStore.
```
```suggestion
   This will cause Cilium to read identities only from CRDs, but continue to write them to the KVStore.
```

learnitall commented on Documentation/operations/upgrade.rst:
```diff
@@ -463,36 +463,48 @@ As the :term:`ConfigMap` is successfully upgraded we can start upgrading Cilium
-command if desired. It will identify already allocated identities or ones that
+There are two ways to achieve this: you can either run a one-off ``cilium preflight migrate-identity`` script
+which will perform a point-in-time copy of all identities from the kvstore to CRDs (added in Cilium 1.6), or use the "Double Write" identity
+allocation mode which will have Cilium manage identities in both the kvstore and CRD at the same time for a seamless migration (added in Cilium 1.17).
```
Is `added in Cilium 1.17` accurate? I think that these changes will land in a patch release of 1.16 before they land in 1.17.

antonipp replied:
I am not sure tbh but Marco suggested 1.17 here: https://github.com/cilium/cilium/pull/31920#discussion_r1660993202

There are still a couple of unresolved issues in this PR so I am not sure how long all of it will take.

giorio94 replied:
> I think that these changes will land in a patch release of 1.16 before they land in 1.17.

Hmm, not sure I'm following. This PR will be part of v1.17, as it missed the v1.16 feature freeze some weeks ago, and being a new feature it doesn't qualify for backport to stable branches.

learnitall replied:
Ah I see, I got confused. Thanks for clarifying!

learnitall commented on Documentation/operations/upgrade.rst:
```diff
@@ -563,12 +575,60 @@ Once the migration is complete, confirm the endpoint identities match by listing
 
+.. _double_write_migration:
+
+Migration with the "Double Write" identity allocation mode
```
The docs here describe a KVStore -> CRD migration, but couldn't this technically also be used for a CRD -> KVStore migration?

antonipp replied:
Yeah, it's possible to do this as well! I added an explicit note about it in the docs.

learnitall commented on Documentation/operations/upgrade.rst:
```diff
@@ -563,12 +575,60 @@ Once the migration is complete, confirm the endpoint identities match by listing
+Rollout Instructions
+~~~~~~~~~~~~~~~~~~~~
+
+#. Re-deploy the Operator and the Agents with ``--identity-allocation-mode=doublewrite-readkvstore``.
```
Is there a specific order to the rollout between the Operator and the Agents?

antonipp replied:
There is no strict requirement but it's actually better to switch the Operator to Double Write mode first and then keep it this way until the Agents are fully migrated. This is because the Operator does GC so you don't want to be in a situation where the Agents write identities but they are not being cleaned up.
I updated the docs to reflect that.

learnitall commented on Documentation/operations/upgrade.rst:
```diff
@@ -563,12 +575,60 @@ Once the migration is complete, confirm the endpoint identities match by listing
+
+   * ``cilium_operator_identity_crd_total_count`` and ``cilium_operator_identity_kvstore_total_count`` reporting the total number of identities in CRDs and KVStore respectively.
+   * ``cilium_operator_identity_crd_only_count`` and ``cilium_operator_identity_kvstore_only_count`` reporting the number of
+     identities that are only in CRDs or KVStore respectively. These metrics should help with detecting inconsistencies.
```
```suggestion
     identities that are only in CRDs or only in the KVStore respectively, to help detect inconsistencies.
```

learnitall reviewed (changes requested):
The doc changes look great! I really appreciate the amount of detail that you put into them. I left some suggestions for grammar and word choice, as well as a couple of questions to help me understand the changes. Let me know what you think, thanks!

squeed commented on pkg/kvstore/allocator/doublewrite/backend.go:
```diff
@@ -0,0 +1,227 @@
+	log = logging.DefaultLogger.WithField(logfields.LogSubsys, "double-write-allocator")
+)
+
+func NewDoubleWriteBackend(c DoubleWriteBackendConfiguration) (allocator.Backend, error) {
```
nit: Could you add a docblock here that gives a quick overview of what this does? Doesn't have to be fancy, just a few sentences would do. Thanks!

antonipp replied:
Done!

squeed commented on pkg/kvstore/allocator/doublewrite/backend.go:
```diff
@@ -0,0 +1,227 @@
+	return d.kvstoreBackend.RunGC(ctx, rateLimit, staleKeysPrevRound, minID, maxID)
+}
+
+func (d *doubleWriteBackend) UpdateKey(ctx context.Context, id idpool.ID, key allocator.AllocatorKey, reliablyMissing bool) error {
```
it seems like `reliablyMissing` isn't really used anymore?

antonipp replied:
Yes, there is actually a thread about it here: https://github.com/cilium/cilium/pull/31920#discussion_r1650610739

giorio94 commented on pkg/kvstore/allocator/doublewrite/backend.go:
```diff
@@ -0,0 +1,267 @@
+	if kvStoreErr != nil {
+		log.WithFields(logrus.Fields{logfields.Identity: id.String(), logfields.Key: key.String()}).WithError(kvStoreErr).Error("KVStore backend failed to allocate identity, deleting the corresponding CRD identity")
+		// revert the allocation in the CRD backend
+		releaseErr := d.crdBackend.DeleteID(ctx, id)
+		if releaseErr != nil {
+			log.WithFields(logrus.Fields{logfields.Identity: id.String(), logfields.Key: crdKey.String()}).WithError(releaseErr).Error("CRD backend failed to release identity")
+		}
+		return kvStoreKey, kvStoreErr
```
This part is tricky, and AFAIU it could open to a possible race condition with the following sequence of operations:

* Identity x is successfully allocated in CRD mode
* The creation event is received by other Cilium agents configured in read-crd mode via the informer, which possibly assign the given identity to a newly started endpoint sharing the same labels.
* Identity x fails to be allocated in kvstore mode, and it gets deleted
* There are now endpoints associated with an identity that doesn't exist anymore.

A similar race condition could happen without reverting the allocation as well though. In that case, the CRD identity the other endpoints are associated with would still be present, although it would be missing from the kvstore. 

Still, performing the revert ensures that the incomplete identity is not subsequently reused, which would be instead possible if not cleaned-up (this sounds more likely to occur than the other race condition mentioned above).

While no solution seems perfect in my opinion (and there's most likely no way to prevent all possible race conditions), I actually tend to prefer the current implementation based on the above considerations.

antonipp replied:
Makes sense, let's leave this as-is then (I agree that hunting all possible race conditions will be a complicated endeavour)

giorio94 commented on pkg/kvstore/allocator/doublewrite/backend.go:
```diff
@@ -0,0 +1,267 @@
+	if crdErr != nil {
+		logEntry := log.WithFields(logrus.Fields{logfields.Identity: id.String(), logfields.Key: key.String()}).WithError(crdErr)
+		logMessage := "CRD backend failed to acquire reference with lock"
+		if strings.Contains(crdErr.Error(), "does not exist") {
```
Nit. I'd skip the log only when `readFromKVStore` is set to true, as only expected in that case.

giorio94 reviewed (approved):
Thanks, looks good to me. Just a couple of final comments inline.

I performed a local test migration, and managed to switch from kvstore to crd without breaking any existing connections, even if matched by ingress/egress network policies :rocket:

ti-mo reviewed (approved):
Approving since I can't unassign myself as a reviewer and I don't own any code here.

learnitall reviewed (approved):
Looks great! Thank you!

pippolo84 commented on operator/doublewrite/metricreporter.go:
```diff
@@ -0,0 +1,193 @@
+}
+
+func (g *DoubleWriteMetricReporter) Stop(ctx cell.HookContext) error {
+	if g.mgr != nil {
```
We should add the same guard as in `Start` here:

```go
if option.Config.IdentityAllocationMode != option.IdentityAllocationModeDoubleWriteReadKVstore && option.Config.IdentityAllocationMode != option.IdentityAllocationModeDoubleWriteReadCRD {
		return nil
	}
```

Otherwise, in case we are not running with the double-write backend we'll panic trying to close the nil channel `g.crdBackendWatcherStopChan`

Even better, we should put that guard at the top of `registerDoubleWriteMetricReporter`, in order to avoid creating `doubleWriteMetricReporter` and registering it to the lifecycle.

antonipp replied:
Makes sense, I just moved it to `registerDoubleWriteMetricReporter`

pippolo84 commented on operator/doublewrite/cell.go:
```diff
@@ -0,0 +1,42 @@
+
+var Cell = cell.Module(
+	"double-write-metric-reporter",
+	"Double-Write Metric Reporter",
```
Nit: ideally this field should be more descriptive, it would be better to write something along the lines of the comment used in `operator/cmd/root.go` when including this cell in the operator hive.

pippolo84 commented on operator/doublewrite/metricreporter.go:
```diff
@@ -0,0 +1,193 @@
+	g.wg = sync.WaitGroup{}
+	g.wg.Add(1)
+	go func() {
+		g.crdBackend.ListAndWatch(context.Background(), NoOpHandlerWithListDone{listDone: g.crdBackendListDone}, g.crdBackendWatcherStopChan)
```
It would be better to create a context with cancellation and save its cancel function in the `DoubleWriteMetricReporter`, so to signal `ListAndWatch` in case `Stop` is called before it is completed.

antonipp replied:
I looked at the implementation of `ListAndWatch` and it seems like it's never even using the `ctx`: https://github.com/cilium/cilium/blob/b8b8d813c3fede25c6b94064356506b944729a03/pkg/k8s/identitybackend/identity.go#L373-L428 

The only way to actually stop the operations is to close the `stopChan` which is what I already do here: https://github.com/cilium/cilium/blob/258384179499c70f13dec1e01bce4b2c1697c045/operator/doublewrite/metricreporter.go#L124

pippolo84 replied:
Ouch. Definitely something to improve on the CRD backend side, but totally outside of the scope of this PR.
Good catch, thanks! :+1:

pippolo84 commented on operator/doublewrite/metricreporter.go:
```diff
@@ -0,0 +1,193 @@
+
+	g.mgr = controller.NewManager()
+	g.mgr.UpdateController("double-write-metric-reporter",
+		controller.ControllerParams{
+			Group:       controller.NewGroup("double-write-metric-reporter"),
+			RunInterval: g.interval,
+			DoFunc:      g.compareCRDAndKVStoreIdentities,
+		})
```
Non-blocking: have you had a look at `hive/job` package? A `job.Timer` can be used here instead, and it gives better integration with hive, health reporting, and other benefits.

However, given the expected usage of this cell, we can leave this as a follow up PR.

antonipp replied:
I didn't know that this existed! But yes, it's a bit of work, I can do it as a follow-up

pippolo84 commented on operator/doublewrite/metricreporter.go:
```diff
@@ -0,0 +1,193 @@
+	// Compare CRD and KVStore identities
+	maxPrintedDiffIDs := 5 // Cap the number of differing IDs so as not to log too many
+	onlyInCrdCount, onlyInCrdSample := difference(crdIdentityIds, kvstoreIdentityIds, maxPrintedDiffIDs)
+	onlyInKVStoreCount, onlyInKVStoreSample := difference(kvstoreIdentityIds, crdIdentityIds, maxPrintedDiffIDs)
```
Alternatively, we can rely on `Diff` from `github.com/cilium/cilium/pkg/slices` package. The only difference with the `difference` implementation above is that we'll allocate memory for all the samples in the diff and then limit it later to `maxPrintedDiffIDs`. Given the frequency this comparison is running, alongside the fact that `idpool.ID` is a `uint64` it shouldn't be much of a problem.
In case you want to go ahead with the current implementation, I suggest to add some unit tests for `difference`.

antonipp replied:
Nice, I didn't know this existed. You are right it's a bit less optimized but should be fine in this case. I removed my implementation and replaced it with `slices.Diff`.

pippolo84 reviewed (changes requested):
Thanks for the PR, great work. :100: 

Left some comments inline, nothing major.

Regarding https://github.com/cilium/cilium/pull/31920#discussion_r1664243771, my understanding is the same as @giorio94 , so I'd say we should proceed with propagating `reliablyMissing` for kvstore backend.

julianwiedmann commented on pkg/egressgateway/manager.go:
```diff
@@ -171,8 +171,10 @@ func NewEgressGatewayManager(p Params) (out struct {
+	if dcfg.IdentityAllocationMode == option.IdentityAllocationModeKVstore ||
+		dcfg.IdentityAllocationMode == option.IdentityAllocationModeDoubleWriteReadKVstore ||
+		dcfg.IdentityAllocationMode == option.IdentityAllocationModeDoubleWriteReadCRD {
+		return out, fmt.Errorf("egress gateway is not supported in %s identity allocation mode", dcfg.IdentityAllocationMode)
```
```suggestion
	if dcfg.IdentityAllocationMode != option.IdentityAllocationModeCRD {
		return out, fmt.Errorf("egress gateway is not supported in %s identity allocation mode", dcfg.IdentityAllocationMode)
```

While addressing Fabio's feedback, could you please also make this part future-proof by explicitly checking for the *one* supported allocation mode? :)

pippolo84 reviewed (approved):
LGTM, thanks! :100:
